### PR TITLE
Added an node e2e test for pod evictions due to disk pressure.

### PR DESCRIPTION
**Pull request by ronnielai**
#29800

### CODE REVIEW

File: test/e2e/framework/test_context.go [hunk 1]
```diff
@@ -150,4 +152,5 @@ func RegisterClusterFlags() {
 func RegisterNodeFlags() {
 	flag.StringVar(&TestContext.NodeName, "node-name", "", "Name of the node to run tests on (node e2e suite only).")
 	flag.BoolVar(&TestContext.CgroupsPerQOS, "cgroups-per-qos", false, "Enable creation of QoS cgroup hierarchy, if true top level QoS and pod cgroups are created.")
+	flag.StringVar(&TestContext.EvictionHard, "eviction-hard", "", "the hard eviction thresholds.")
```
Comment (Random-Liu) [inline test/e2e/framework/test_context.go]: s/the/The. A short description of what this flag is for, and give a simple example.

File: test/e2e_node/disk_eviction_test.go [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+	"k8s.io/kubernetes/test/e2e/framework"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+)
+
+const (
+	// The interval seconds between pod status checks
+	POD_CHECK_INTERVAL = 2
```
Comment (Random-Liu) [inline test/e2e_node/disk_eviction_test.go]: s/POD_CHECK_INTERVAL/podCheckInterval
https://github.com/golang/go/wiki/CodeReviewComments#mixed-caps

File: test/e2e_node/disk_eviction_test.go [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+	"k8s.io/kubernetes/pkg/util/uuid"
+	"k8s.io/kubernetes/test/e2e/framework"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+)
+
+const (
+	// The interval seconds between pod status checks
```
Comment (Random-Liu) [inline test/e2e_node/disk_eviction_test.go]: nit: podCheckInterval is the interval seconds between pod status checks.
https://github.com/golang/go/wiki/CodeReviewComments#comment-sentences

File: test/e2e_node/disk_eviction_test.go [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+	POD_CHECK_INTERVAL = 2
+
+	DUMMY_FILE = "dummy."
```
Comment (Random-Liu) [inline test/e2e_node/disk_eviction_test.go]: dito

File: test/e2e_node/disk_eviction_test.go [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+		c = f.Client
+		nodeList := framework.GetReadySchedulableNodesOrDie(c)
+		n = &nodeList.Items[0]
+	})
+
+	Describe("hard eviction test", func() {
+		Context("pod using the most disk space gets evicted when the node disk usage is above the eviction hard threshold", func() {
+			var busyPodName, idlePodName string
+			var containersToCleanUp map[string]bool
+			evictionOccurred := false
```
Comment (Random-Liu) [inline test/e2e_node/disk_eviction_test.go]: Why put `evictionOccurred` here? :)

File: test/e2e_node/disk_eviction_test.go (hunk 2 above)
Comment (ronnielai) [inline test/e2e_node/disk_eviction_test.go]: done

File: test/e2e_node/disk_eviction_test.go (hunk 3 above)
Comment (ronnielai) [inline test/e2e_node/disk_eviction_test.go]: done

File: test/e2e_node/disk_eviction_test.go (hunk 4 above)
Comment (ronnielai) [inline test/e2e_node/disk_eviction_test.go]: done

File: test/e2e_node/disk_eviction_test.go (hunk 5 above)
Comment (ronnielai) [inline test/e2e_node/disk_eviction_test.go]: moved.

File: test/e2e_node/disk_eviction_test.go [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+	"k8s.io/kubernetes/test/e2e/framework"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+)
+
+const (
+	// podCheckInterval is the interval seconds between pod status checks.
+	podCheckInterval = 2
```
Comment (Random-Liu) [inline test/e2e_node/disk_eviction_test.go]: Use `2 * time.Second` directly here?

File: test/e2e_node/disk_eviction_test.go [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+										dummyFile)},
+							},
+						},
+					},
+				})
+			})
+
+			It("should evict the pod using the most disk space", func() {
+				if !evictionOptionIsSet() {
+					framework.Logf("test skipped")
```
Comment (Random-Liu) [inline test/e2e_node/disk_eviction_test.go]: nit: test skipped because eviction option is not set

File: test/e2e_node/disk_eviction_test.go [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+					return nil
+				}, time.Minute*5, time.Second*podCheckInterval).Should(BeNil())
+			})
+		})
+	})
+})
+
+func verifyPodEviction(podData *api.Pod) error {
+	if podData.Status.Phase != "Failed" {
+		return fmt.Errorf("expected phase to be failed. Got %+v", podData.Status.Phase)
```
Comment (Random-Liu) [inline test/e2e_node/disk_eviction_test.go]: nit: s/Got/got

File: test/e2e_node/disk_eviction_test.go [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+	}
+	if podData.Status.Reason != "Evicted" {
+		return fmt.Errorf("expected state reason to be evicted. Got %+v", podData.Status.Reason)
```
Comment (Random-Liu) [inline test/e2e_node/disk_eviction_test.go]: ditto

File: test/e2e_node/disk_eviction_test.go [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+	}
+	return nil
+}
+
+func nodeHasDiskPressure(c *client.Client) bool {
+	nodeList := framework.GetReadySchedulableNodesOrDie(c)
+	for _, condition := range nodeList.Items[0].Status.Conditions {
+		if condition.Type == "DiskPressure" {
```
Comment (Random-Liu) [inline test/e2e_node/disk_eviction_test.go]: Use [`api.NodeDiskPressure`](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/types.go#L2073)

File: test/e2e_node/disk_eviction_test.go [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+			return condition.Status == "True"
```
Comment (Random-Liu) [inline test/e2e_node/disk_eviction_test.go]: Use [api.ConditionTrue](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/types.go#L1048)

**Comment by Random-Liu** [comment]
LGTM with nits.

File: test/e2e_node/disk_eviction_test.go (hunk 6 above)
Comment (ronnielai) [inline test/e2e_node/disk_eviction_test.go]: done

File: test/e2e_node/disk_eviction_test.go (hunk 7 above)
Comment (ronnielai) [inline test/e2e_node/disk_eviction_test.go]: done

File: test/e2e_node/disk_eviction_test.go (hunk 8 above)
Comment (ronnielai) [inline test/e2e_node/disk_eviction_test.go]: done

File: test/e2e_node/disk_eviction_test.go (hunk 9 above)
Comment (ronnielai) [inline test/e2e_node/disk_eviction_test.go]: done

File: test/e2e_node/disk_eviction_test.go (hunk 10 above)
Comment (ronnielai) [inline test/e2e_node/disk_eviction_test.go]: done

File: test/e2e_node/disk_eviction_test.go (hunk 11 above)
Comment (ronnielai) [inline test/e2e_node/disk_eviction_test.go]: done

**Comment by Random-Liu** [comment]
LGTM.

**Pull request merged**